### PR TITLE
Osquerybeat: Update ECS mapping implementation based on the latest developers feedback

### PR DESCRIPTION
**Pull request by aleksmaus**
## What does this PR do?
Changes the format of ECS mapping configuration for the queries:
1. The mapping keys are the ECS key path
2. The mapping value is an object that allows to define the static value or the field which is the name of the osquery result column.

Improves test coverage for configuration plugin.

Current format example:
```
                        "ecs_mapping": {
                                "user.custom.shoeSize": {
                                    "value": 45
                                },
                                "user.id": {
                                    "field": "uid"
                                },
                                "user.name": {
                                    "field": "username"
                                }
                        }
```

## Why is it important?

ECS mapping support for Osquerybeat

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas

## Related issues

- Requires https://github.com/elastic/integrations/pull/1360

## Screenshots

The fragment of the document showing and example of the ECS fields mapped:

Screen Shot 2021-07-27 at 3 19 42 PM

### CODE REVIEW

File: x-pack/osquerybeat/beater/osquerybeat.go [hunk 1]
```diff
@@ -278,7 +278,11 @@ func (bt *osquerybeat) runOsquery(ctx context.Context, b *beat.Beat, osq *osqd.O
 				bt.log.Info("context cancelled, exiting")
 				return ctx.Err()
 			case inputConfigs := <-inputCh:
-				configPlugin.Set(inputConfigs)
+				err = configPlugin.Set(inputConfigs)
+				if err != nil {
+					bt.log.Errorf("failed to set configuration from inputs: %v", err)
+					return err
```
Comment (aleksmaus) [inline x-pack/osquerybeat/beater/osquerybeat.go:284]: actually not sure if we should fail to run osquerybeat, in this case in particular because of bad ecs mapping or maybe we should just log the invalid/rejected mapping and continue running. thoughts?

File: x-pack/osquerybeat/beater/config_plugin.go [hunk 2]
```diff
@@ -190,13 +195,22 @@ func flattenECSMapping(m map[string]interface{}) (ecs.Mapping, error) {
...
 		}
 		return nil
-	} else if path[len(path)-1] == "value" {
+	} else if path[len(path)-1] == keyValue {
+		if len(path) == 1 {
+			return fmt.Errorf("unexpected top level key '%s': %w", keyValue, ErrECSMappingIsInvalid)
+		}
+		if len(path) == 1 {
+			return ErrECSMappingIsInvalid
+		}
```
Comment (urso) [inline x-pack/osquerybeat/beater/config_plugin.go]: nit; Noop branch, we already know `len(path) > 1` when we reach the if statement.

File: x-pack/osquerybeat/beater/config_plugin_test.go [hunk 3]
```diff
@@ -116,13 +117,45 @@ func TestFlattenECSMapping(t *testing.T) {
...
+		newmap := make(map[string]interface{})
+		m[string(key)] = newmap
+		m = newmap
+		key += 1
+	}
+	m[k] = v
+	return res
+}
+
+func TestFlattenECSMappingEdges(t *testing.T) {
```
Comment (urso) [inline x-pack/osquerybeat/beater/config_plugin_test.go:135]: do we also need a test for the error case?

File: x-pack/osquerybeat/beater/config_plugin.go (hunk 2 above)
Comment (aleksmaus) [reply x-pack/osquerybeat/beater/config_plugin.go]: yeah, that was updated in the subsequent commit in this PR

File: x-pack/osquerybeat/beater/config_plugin_test.go (hunk 3 above)
Comment (aleksmaus) [reply x-pack/osquerybeat/beater/config_plugin_test.go:135]: this test covers some error cases, will do another pass to double check what other cases can be added

File: x-pack/osquerybeat/beater/config_plugin.go [hunk 4]
```diff
@@ -17,6 +20,15 @@ import (
 const (
 	configName           = "osq_config"
 	scheduleSplayPercent = 10
+	maxECSMappingDepth   = 25 // max ECS dot delimited key path
```
Comment (scunningham) [inline x-pack/osquerybeat/beater/config_plugin.go:23]: Arbitrary depth? This one goes to 11?

**Review by scunningham — APPROVED** [review]
Looks ok.  I would suggest going over tests again to make sure you've got all the unhappy path edge cases: 

- empty string
- empty string with whitepaces
- empty fields 
- empty fields with whitespace
- empty field at end with or w/o whitepace
- crazy unicode values or escaped dots
- crazy large values, 1meg of huge fields.
- crazy small values at max depth + 1

Just some thoughts on how to break it.

File: x-pack/osquerybeat/beater/config_plugin.go (hunk 4 above)
Comment (aleksmaus) [reply x-pack/osquerybeat/beater/config_plugin.go]: extended the comment

**Comment by aleksmaus** [comment]
> Looks ok. I would suggest going over tests again to make sure you've got all the unhappy path edge cases:
> 
> * empty string
> * empty string with whitepaces
> * empty fields
> * empty fields with whitespace
> * empty field at end with or w/o whitepace
> * crazy unicode values or escaped dots
> * crazy large values, 1meg of huge fields.
> * crazy small values at max depth + 1
> 
> Just some thoughts on how to break it.

added more edge cases handling and test coverage
